src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs

```diff
@@ -0,0 +1,112 @@
+            | Foo(out x.Prop) |      |       |             |     ✔️      |
+            | Foo(ref x.Prop) |      |       |     ✔️      |     ✔️      |
+
+            * where void Foo(in T v)
+            */
+
+            if (operation.Parent is IAssignmentOperation assignmentOperation &&
+                assignmentOperation.Target == operation)
+            {
+                return operation.Parent.Kind == OperationKind.CompoundAssignment
+                    ?  ValueUsageInfo.ReadWrite
+                    : ValueUsageInfo.Write;
+            }
+            else if (operation.Parent is IIncrementOrDecrementOperation)
+            {
+                return ValueUsageInfo.ReadWrite;
+            }
+            else if (operation.Parent is IParenthesizedOperation parenthesizedOperation)
+            {
+                return parenthesizedOperation.GetValueUsageInfo();
+            }
+            else if (operation.Parent is INameOfOperation ||
+                operation.Parent is ITypeOfOperation ||
+                operation.Parent is ISizeOfOperation)
+            {
+                return ValueUsageInfo.None;
+            }
+            else if (operation.Parent is IArgumentOperation argumentOperation)
+            {
+                switch (argumentOperation.Parameter.RefKind)
+                {
+                    case RefKind.RefReadOnly:
+                        return ValueUsageInfo.ReadableRef;
+
+                    case RefKind.Out:
+                        return ValueUsageInfo.WritableRef;
+
+                    case RefKind.Ref:
+                        return ValueUsageInfo.ReadableWritableRef;
+
+                    default:
+                        return ValueUsageInfo.Read;
+                }
+            }
+            else if (IsInLeftOfDeconstructionAssignment(operation))
+            {
+                return ValueUsageInfo.Write;
+            }
+
+            return ValueUsageInfo.Read;
+        }
+
+        private static bool IsInLeftOfDeconstructionAssignment(IOperation operation)
+        {
+            var previousOperation = operation;
+            operation = operation.Parent;
+
+            while (operation != null)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.DeconstructionAssignment:
+                        var deconstructionAssignment = (IDeconstructionAssignmentOperation)operation;
+                        return deconstructionAssignment.Target == previousOperation;
+
+                    case OperationKind.Tuple:
+                    case OperationKind.Conversion:
+                    case OperationKind.Parenthesized:
+                        previousOperation = operation;
+                        operation = operation.Parent;
+                        continue;
+
+                    default:
+                        return false;
+                }
+            }
+
+            return false;
```

Comment:
tagging @jcouv

Reply:
was this resolved by @jcouv confirming the logic here? Thanks!

Reply:
@jcouv does this seem correct to you?

---
In reply to: [213420369](https://github.com/dotnet/roslyn/pull/29511#discussion_r213420369) [](ancestors = 213420369)